Refactor connectivity netsocket unittests

hazzlim opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
Clean up CMake definitions of connectivity-netsocket unit tests to only include required source files, and to depend only on headers each test uses.

Fixes #14879

#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->
None.
----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->
@rajkan01 @ARMmbed/mbed-os-core
----------------------------------------------------------------------------------------------------------------

hazzlim committed:
Unittests: Make netsocket test header dependencies explicit

The CMake target `mbed-headers` brings in all
headers, and we are gradually moving away from it and explicitly use
only headers needed by each unit test.

hazzlim committed:
Unittests: connectivity: remove trailing whitespace

hazzlim committed:
Unittests: Clean up unused sources from netsocket

Some source files included when building netsocket unit tests with CMake
were determined to be unused by removing the files from target_sources()
and verifying that the tests still built and ran successfully. These
files have been removed from CMake builds.

rwalton-arm commented on connectivity/netsocket/tests/UNITTESTS/netsocket/DTLSSocket/CMakeLists.txt:
```diff
@@ -12,19 +12,12 @@ target_compile_definitions(${TEST_NAME}
 
 target_sources(${TEST_NAME}
     PRIVATE
-        ${mbed-os_SOURCE_DIR}/connectivity/netsocket/source/SocketAddress.cpp
```
In the commit "Unittests: Clean up unused sources from netsocket" how did we determine the files weren't used? Please add the reasoning to the commit message. 

It also looks like we added linking to `mbed-stubs-nanostack-libservice`. Did we move the "unused sources" to that library in some other commit/PR, or are they actually unused? I can't tell from the commit description.

hazzlim replied:
The files were determined to be unused by removing them from `target_sources` and verifying that the test still compiled and ran successfully. I understood from speaking to Rajkumar that this was sufficient reasoning, I will add the reasoning to the commit message description. 

The linking to `mbed-stubs-nanostack-libservice` was to use the stubs of `stoip4.c` and `ip4tos.c` instead of including those source files. I can also amend the commit message to explain this, or would it be more appropriate to move these changes to a separate commit ?

hazzlim committed:
Unittests: Replace files with stubs in netsocket unit tests

Some netsocket unit tests depend on nanostack-libservice library, but
should use stubs instead of including the source files. We remove the
source files and link to mbed-stubs-nanostack-libservice.

rwalton-arm reviewed (dismissed):
LGTM. Thanks for the changes.
